File: ios/Sources/KlinikTravelFeature/TravelModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum TravelPhase: Sendable, Equatable {
    case loading
    case loaded
    case notFound
    case failed(String)
}

public struct TravelState: Sendable, Equatable {
    public var phase: TravelPhase = .loading
    public var plan: TravelPlan?
    public var clearedToFlyBy: String?
    public var saving = false
    public var error: String?

    public init() {}
}

/**
 * A patient's trip (spec M19).
 *
 * Two callers with different rights. A coordinator fills the whole thing in
 * with `patients.write` and never sees a lab result; a doctor additionally says
 * whether the patient may fly, which needs `medical.decide`. The model does not
 * enforce either — the server does — but it does render the difference, so
 * nobody is shown a switch they will be refused.
 */
@MainActor
public final class TravelModel {
    private let api: TravelAPI
    /// Nil on the patient's own copy, which reads `me/travel` and edits nothing.
    private let patientId: String?

    private var state = TravelState()

    public init(api: TravelAPI, patientId: String? = nil) {
        self.api = api
        self.patientId = patientId
    }

    public func currentState() -> TravelState { state }

    public func load() async {
        do {
            let view = patientId.map { id in { try await self.api.forPatient(id) } }
                ?? { try await self.api.mine() }

            let loaded = try await view()
            state.plan = loaded.plan
            state.clearedToFlyBy = loaded.clearedToFlyBy
            state.phase = .loaded
        } catch APIError.notFound {
            state.phase = .notFound
        } catch let error as APIError {
            state.phase = .failed(L10n.message(for: error))
        } catch {
            state.phase = .failed(L10n.string("error.server"))
        }
    }

    public func save(_ edit: TravelPlanEdit) async -> Bool {
        guard let patientId else { return false }

        state.saving = true
        state.error = nil

        defer { state.saving = false }

        do {
            var payload = edit
            payload.expectedVersion = state.plan?.version

            state.plan = try await api.save(patientId: patientId, payload)

            return true
        } catch APIError.conflict {
            state.error = L10n.string("file.versionConflict")
        } catch let error as APIError {
            state.error = L10n.message(for: error)
        } catch {
            state.error = L10n.string("error.server")
        }

        return false
    }

    /// A clinician's decision, not a calculation.
    public func setClearedToFly(_ cleared: Bool) async {
        guard let patientId else { return }

        state.saving = true
        state.error = nil

        defer { state.saving = false }

        do {
            state.plan = try await api.setClearedToFly(patientId: patientId, cleared: cleared)
            // Re-read for the name of whoever just signed it off; the write
            // returns the plan and not the person.
            await load()
        } catch APIError.forbidden {
            // A coordinator pressing this is not an error worth a red banner —
            // they simply may not, and the screen should say which.
            state.error = L10n.string("travel.clearanceForbidden")
        } catch let error as APIError {
            state.error = L10n.message(for: error)
        } catch {
            state.error = L10n.string("error.server")
        }
    }
}

File: android/feature/travel/src/main/kotlin/xyz/klinik/feature/travel/TravelModel.kt
package xyz.klinik.feature.travel

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import xyz.klinik.network.ApiError
import xyz.klinik.network.TravelApi
import xyz.klinik.network.TravelPlan
import xyz.klinik.network.TravelPlanView
import xyz.klinik.network.UiText
import xyz.klinik.network.UpsertTravelPlan
import xyz.klinik.network.uiText

sealed interface TravelPhase {
    data object Loading : TravelPhase
    data object Loaded : TravelPhase

    /** Nothing entered yet. A real state, and different from a failure. */
    data object None : TravelPhase
    data class Failed(val message: UiText) : TravelPhase
}

data class TravelState(
    val phase: TravelPhase = TravelPhase.Loading,
    val view: TravelPlanView? = null,
    val draft: UpsertTravelPlan = UpsertTravelPlan(),
    val editing: Boolean = false,
    val busy: Boolean = false,
    val error: UiText? = null,
) {
    val plan: TravelPlan? get() = view?.plan
    val isClearedToFly: Boolean get() = plan?.isClearedToFly == true
    val clearedBy: String? get() = view?.clearedToFlyBy
}

/**
 * Getting the patient here and home again (spec M14).
 *
 * Almost all of this is logistics a coordinator owns — a flight, a hotel,
 * somebody at the airport. Exactly one field is not: whether a clinician has
 * said the patient may fly. It is saved through its own call rather than with
 * the rest of the form, so a coordinator editing a hotel address can never
 * carry a medical decision along with it.
 */
class TravelModel(
    private val api: TravelApi,
    /** Null on the patient's own screen, which is read-only. */
    private val patientId: String? = null,
) {
    private val _state = MutableStateFlow(TravelState())
    val state: StateFlow<TravelState> = _state.asStateFlow()

    suspend fun load() {
        _state.value = _state.value.copy(phase = TravelPhase.Loading)

        try {
            val view = if (patientId == null) api.mine() else api.forPatient(patientId)
            val plan = view.plan

            _state.value = TravelState(
                phase = if (plan == null || plan.isEmpty) TravelPhase.None else TravelPhase.Loaded,
                view = view,
                draft = plan.toDraft(),
            )
        } catch (error: Throwable) {
            _state.value = _state.value.copy(phase = TravelPhase.Failed(messageFor(error)))
        }
    }

    fun beginEditing() {
        _state.value = _state.value.copy(
            editing = true,
            draft = _state.value.plan.toDraft(),
            error = null,
        )
    }

    fun cancelEditing() {
        _state.value = _state.value.copy(editing = false, error = null)
    }

    fun edit(change: UpsertTravelPlan.() -> UpsertTravelPlan) {
        _state.value = _state.value.copy(draft = _state.value.draft.change())
    }

    suspend fun save(): Boolean {
        val target = patientId ?: return false

        _state.value = _state.value.copy(busy = true, error = null)

        return try {
            val view = api.save(target, _state.value.draft)

            _state.value = _state.value.copy(
                phase = if (view.plan?.isEmpty != false) TravelPhase.None else TravelPhase.Loaded,
                view = view,
                draft = view.plan.toDraft(),
                editing = false,
                busy = false,
            )

            true
        } catch (error: Throwable) {
            // The form stays open with what was typed. A conflict here means
            // somebody else saved first, and throwing the entry away would
            // lose the second coordinator's work as well as the first's.
            _state.value = _state.value.copy(busy = false, error = messageFor(error))
            false
        }
    }

    /**
     * A clinician's signature on the flight home.
     *
     * Its own call, and a coordinator without the permission is told so by the
     * server rather than being offered a switch that silently does nothing.
     */
    suspend fun setClearedToFly(cleared: Boolean): Boolean {
        val target = patientId ?: return false

        _state.value = _state.value.copy(busy = true, error = null)

        return try {
            val view = api.setClearedToFly(target, cleared)

            _state.value = _state.value.copy(view = view, busy = false)
            true
        } catch (error: Throwable) {
            _state.value = _state.value.copy(busy = false, error = messageFor(error))
            false
        }
    }

    private fun messageFor(error: Throwable): UiText =
        (error as? ApiError)?.uiText() ?: UiText.Key("error.server")
}

/**
 * The saved plan as a form.
 *
 * `expectedVersion` travels with it: two coordinators arranging one patient's
 * travel is the ordinary case here, and a second save winning silently would
 * lose the first's flight number.
 */
private fun TravelPlan?.toDraft(): UpsertTravelPlan = UpsertTravelPlan(
    arrivalFlight = this?.arrivalFlight,
    arrivalAt = this?.arrivalAt,
    departureFlight = this?.departureFlight,
    departureAt = this?.departureAt,
    hotelName = this?.hotelName,
    hotelAddress = this?.hotelAddress,
    hotelCheckIn = this?.hotelCheckIn,
    hotelCheckOut = this?.hotelCheckOut,
    greeterName = this?.greeterName,
    greeterPhone = this?.greeterPhone,
    transferNote = this?.transferNote,
    interpreterName = this?.interpreterName,
    interpreterLanguage = this?.interpreterLanguage,
    interpreterPhone = this?.interpreterPhone,
    notes = this?.notes,
    expectedVersion = this?.version,
)
